### PR TITLE
[velux] fix concurrency bugs, other minor issues, update readme.md (replaces #8493)

### PR DESCRIPTION
**Pull request by andrewfg**
**This PR replaces #8493 which got lost in migration**

The binding would not support multiple KLF bridges concurrently.
This PR fixes that problem and a few other minor bugs.
Also updated the ReadMe.md file.
Work-in-Progress => needs review by (author) Guenther Schreiner!

### CODE REVIEW

**Comment by andrewfg** [comment]
@gs4711 new PR created in OH3 to replace the lost / closed #8493

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ public String toString() {
         UNDEFTYPE((short) -1, "Unknown command."),
         // Special item: Shutdown of the connection
         GW_OPENHAB_CLOSE((short) -2, "openHAB connection shutdown command."),
-        // Special item: Shutdown of the connection
+        // Special item: ???
```
Comment (gs4711) [inline bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java:132]: Empty Command (used to send nothing and process an incoming message only).

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/VeluxBindingHandler.java [hunk 2]
```diff
@@ -82,9 +84,16 @@ public VeluxBindingHandler(Thing thing, final Localization localization) {
...
-        return channelUID.getThingUID().getThingTypeUID();
+        Channel chan = getThing().getChannel(channelUID);
+        if (chan != null) {
+            ChannelTypeUID type = chan.getChannelTypeUID();
+            if (type != null) {
+                return new ThingTypeUID(type.getBindingId(), type.getId());
+            }
+        }
+        logger.warn("thingTypeUIDOf({}) failed.", channelUID);
+        return new ThingTypeUID(VeluxBindingConstants.BINDING_ID, "???");
```
Comment (gs4711) [inline bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/VeluxBindingHandler.java:96]: would suggest to replace the three questionmarks by a global constant.

**Review by gs4711 — APPROVED** [review]
The multi-bridge support looks fine to me.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java (hunk 1 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java]: Thank you. Suggestion adopted.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/VeluxBindingHandler.java (hunk 2 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/VeluxBindingHandler.java]: Thank you. Suggestion adopted.

**Comment by andrewfg** [comment]
> The multi-bridge support looks fine to me.

Thanks for the feedback @gs4711. I am getting quite a few IO read timeout exceptions (in spite of increasing the timeout period).

I don't know (yet) if these timeouts are due to my LAN or due to the changes that I made in the code: My original KLF200 is hard wired via Ethernet, but the second (new) KLF is located in an outbuilding, so it is connected via a Wifi bridge comprising two Ubiquiti WiFi APs back-to-back. Therefore the timeouts could be due to the WiFi bridge going down, or due to my code changes. The timeouts always occur at an unexpected time between 2:00 ... 3:00 AM which is irritating since I can never observe them live. 

So I will do the following..
- Modify the logging code to make it clear which KLF is suffering the problem
- Run tests for a few more days
- Determine if the timeouts occur only on the WiFi bridge connected unit
- See if I can improve the error recovery code

**=> Please don't merge until I report back!**

**Comment by andrewfg** [comment]
This PR currently fails the Travis CI test with the message below. I guess that this is because the auto migration of the binding code from v2.5.x to v3.0.0 fails to build in v3.0.0. The reason is that it contains code that was "deprecated" in v2.5.x and completely deleted in v3.0.0. And I suppose that the maintainers removed `org.openhab.binding.velux` from the reactor so that other Travis builds will not fail. Or??

```
[ERROR] [ERROR] Could not find the selected project in the reactor: :org.openhab.binding.velux @ 
[ERROR] Could not find the selected project in the reactor: :org.openhab.binding.velux -> [Help 1]
```
In any case, in this PR there is a [commit](https://github.com/openhab/openhab-addons/pull/8520/commits/090dcf1dc85f12ba4f5ef5dc787cc94e73406625) of a change that does allow the binding to build in v3.0.0. So merging this PR, when it is ready, will also allow `org.openhab.binding.velux` to be added back to the reactor..

**Comment by andrewfg** [comment]
>  I am getting quite a few IO read timeout exceptions ..

Hi @gs4711 I ran it overnight, and can confirm that the IO timeouts do only occur on the KLF unit that is connected via the WiFi bridge. So the strong hypothesis is that they are WiFi dropouts not caused by my code changes.

Nevertheless, I do observe that when a such dropout occurs, instead of simply setting the affected Things' status to offline, the binding is instead "getting its knickers in a twist" and mixing up wrongly matched pairs of GW_XXX_REQ requests and GW_XXX_CFM / GW_XXX_NTF responses; it's an avalanche that can continue for some time before everything stabilises again (or not). This may be due to a problem on the binding client side, which we could potentially fix *), or one on the KLF server side, which we could not. => So I am going to try some more experiments to find a cleaner **) form of escaping from errors.

*) e.g. one idea is: a read time out re-tries 5 times before it finally errors out, which takes 8 or more seconds, and during that time the thread scheduler queue is being filled up with additional polling calls, which may then be executed out of order or too fast; or something like that.. 

**) e.g. one idea is: in case of a comms error the binding make a side channel ping request on the KLF, and if that fails it would ask the binding to cleanly reset its SLIP TLS socket..

File: bundles/pom.xml [hunk 3]
```diff
@@ -280,6 +280,7 @@
     <module>org.openhab.binding.valloxmv</module>
     <module>org.openhab.binding.vektiva</module>
     <module>org.openhab.binding.velbus</module>
+    <module>org.openhab.binding.velux</module>
```
Comment (kaikreuzer) [inline bundles/pom.xml:283]: Ah, much better ;-)

**Comment by andrewfg** [comment]
**Update regarding IO timeout errors**

I think I found the problem. The DataOutputStream and DataInputStream are buffered I/O's. The SLIP protocol requires that prior requests should have been fully sent over the socket, and their responses should have been fully read from the buffer before the next SLIP request is initiated. i.e. in normal operation both read and write buffers should already be empty before the binding initiates the next SLIP call. However in the case that a socket error had occurred, the read buffer often still contains "orphan" data from prior SLIP calls. And furthermore in such a case sending a new SLIP command would add the new command into the write buffer where the socket that had already thrown an exception, meaning that the new SLIP command would not throw a new exception (it is necessary to force the write buffer to actually write to the socket in order to see if another socket exception is thrown). So in the latest commit, I am now doing both of the following things..

- Flushing from the read buffer any "orphan" response data that may have been left over from prior SLIP transactions, and
- Flushing the write buffer directly to the socket to ensure that any exceptions are (re)thrown immediately

I am going to let this run in test on my live system for a day or two, and keep you informed about progress..

**Comment by andrewfg** [comment]
@gs4711 a question to you:

While debugging this PR I wanted to test how the new code handles various kinds of errors. So one of the tests was to see what happened if the KLF is rebooted. Currently the only way to reboot the KLF is to physically power cycle it. However I notice that the KLP SLIP API does contain a command `GW_REBOOT_REQ` which if implemented would potentially allow remote rebooting. So I was wondering if I should perhaps add a new (advanced level) Switch Channel on the Bridge Thing for "Reboot" to this PR?

Note: You do already have an advanced level Switch Channel for Bridge "Reload" (which reloads the binding), so "Reboot" would be very analogous to that.

=> @gs4711 - what do you think?

**Comment by andrewfg** [comment]
I have finished my testing, and all ready for your code review.

**Review by kaikreuzer — APPROVED** [review]
lgtm, thank you!
Let's get that back into openHAB 3 👍

**Comment by andrewfg** [comment]
Back ported to #8702

**Pull request merged**